src/app/src/test/kotlin/org/imperial/mrc/hint/helpers/JSONValidator.kt

```diff
@@ -0,0 +1,85 @@
+package org.imperial.mrc.hint.helpers
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.kotlin.readValue
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Assertions
+import org.leadpony.justify.api.JsonSchema
+import org.leadpony.justify.api.JsonValidationService
+import org.leadpony.justify.api.Problem
+import org.leadpony.justify.api.ProblemHandler
+import java.net.URI
+import java.net.URL
+import javax.json.stream.JsonParsingException
+
+
+class JSONValidator {
+
+    private val service = JsonValidationService.newInstance()
+    private val objectMapper = ObjectMapper()
+
+    private val readerFactory = service.createSchemaReaderFactoryBuilder()
+            .withSchemaResolver(this::resolveSchema)
+            .build()
+
+    private val responseSchema = getSchema("Response")
+
+    fun validateError(response: String, expectedErrorCode: String, expectedErrorMessage: String? = null) {
+        assertValidates(response, responseSchema, "Response")
+        val error = objectMapper.readValue<JsonNode>(response)["errors"].first()
+        val status = objectMapper.readValue<JsonNode>(response)["status"].textValue()
+
+        assertThat(status).isEqualTo("failure")
+        assertThat(error["error"].asText()).isEqualTo(expectedErrorCode)
+        if (expectedErrorMessage != null) {
+            assertThat(error["detail"].asText()).isEqualTo(expectedErrorMessage)
+
+        }
+    }
+
+    fun validateSuccess(response: String, schemaName: String) {
+        assertValidates(response, responseSchema, "Response")
+        val data = objectMapper.readValue<JsonNode>(response)["data"]
+        val status = objectMapper.readValue<JsonNode>(response)["status"].textValue()
+
+        assertThat(status).isEqualTo("success")
+        val dataSchema = getSchema(schemaName)
+        assertValidates(objectMapper.writeValueAsString(data), dataSchema, schemaName)
+    }
+
+    private fun assertValidates(jsonString: String, schema: JsonSchema, schemaName: String) {
+        val problems = mutableListOf<Problem>()
+        val handler = ProblemHandler.collectingTo(problems)
+        val parser = service.createParser(jsonString.byteInputStream(), schema, handler)
+        while (parser.hasNext()) {
+            parser.next()
+        }
+        if (problems.any()) {
+            Assertions.fail<Any>(
+                    "JSON failed schema validation. Attempted to validate: $jsonString against $schemaName."
+            )
+        }
+    }
+
+    private fun getSchema(name: String): JsonSchema {
+        val path = if (name.endsWith(".schema.json")) {
+            name
+        } else {
+            "$name.schema.json"
+        }
+        val url = URL("https://raw.githubusercontent.com/mrc-ide/hintr/mrc-418-rgf/inst/schema/$path")
```

Comment:
It's a bit of a shame the hintr branch has to be configured here as well as in the run-dependencies script - this might be something that comes up a lot during development. Would it be worth doing something like webapps do and keep versions of dependencies to use in text files?